src/Wrapper/Variation.php

```diff
@@ -420,21 +431,43 @@ private function shouldProcessTag(Tag $tag, int $storeId): bool
     private function processImages(): void
     {
         $images = array_merge($this->variationEntity->getImages(), $this->variationEntity->getBase()->getImages());
-
+        $variationAttributes = explode('/', $this->variationAttributesKey);
+        $defaultWrapModeImage = null;
         // Sort images by position.
         usort($images, fn(PimImage $a, PimImage $b) => $a->getPosition() <=> $b->getPosition());
 
-        /** @var ItemImage $image */
         foreach ($images as $image) {
             $imageAvailabilities = $image->getAvailabilities();
             foreach ($imageAvailabilities as $imageAvailability) {
-                if ($imageAvailability->getType() === Availability::STORE) {
+                if ($imageAvailability->getType() !== Availability::STORE) {
+                    continue;
+                }
+
+                if ($this->wrapMode === Product::WRAP_MODE_DEFAULT) {
                     $this->image = new Image($image->getUrlMiddle());
 
                     return;
                 }
+
+                if (!$defaultWrapModeImage) {
+                    $defaultWrapModeImage = $image;
+                }
+
+                $separatedVariation = new SeparatedVariation($this->variationEntity, $this->registryService);
+
+                if (!$separatedVariation->isImageAvailable($image->getAttributeValueImages(), $variationAttributes)) {
+                    continue;
+                }
+
+                $this->image = new Image($image->getUrlMiddle());
+
+                return;
```

Comment:
Those returns after setting the image instances seem a bit wrong. Maybe you want to add a method, which is responsible for getting the image. `processImages` would then be simply responsible for setting it to this class property.

```suggestion
    private function processImages(): void
    {
        $this->image = $this->getVariantImage();
    }
```

Reply:
We need to stop the execution when image is set to an object

Reply:
Yes this is why I would like this to be a method that returns it instead of setting it and returning and therefore stopping further execution. Maybe this example makes it a bit more clear :D

```php
private function processImages(): void
{
    $this->image = $this->getVariantImage();
}

private function getVariantImage(): Image
{
    // foreach ...
    if ($this->wrapMode === Product::WRAP_MODE_DEFAULT) {
        return new Image($image->getUrlMiddle());
    }
    // ...
}
```